File: src/main/java/qmech/helper/objects/BlockBase.java
package qmech.helper.objects;

import net.minecraft.block.ITileEntityProvider;
import net.minecraft.entity.EntityLivingBase;
import net.minecraft.item.ItemStack;
import net.minecraft.util.MathHelper;
import qmech.helper.LoggingHelper;
import qmech.mod.Reference;
import net.minecraft.block.Block;
import net.minecraft.block.material.Material;
import net.minecraft.creativetab.CreativeTabs;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.tileentity.TileEntity;
import net.minecraft.world.World;
import cpw.mods.fml.common.registry.GameRegistry;

public class BlockBase extends Block {

    protected static LoggingHelper logger = LoggingHelper.getInstance();
	
	public static BlockBase config
		(BlockBase block, CreativeTabs ctab, float hardness, float blastResistance, String toolType, int toolLevel) {
        logger.info(String.format("Configuring block (%s) with : \n" +
                ">>> CreativeTab : %s \n" +
                ">>> Hardness : %s, BlastResistance : %s \n" +
                ">>> Tool : %s, %s",
                block.getUnlocalizedName(),
                ctab, hardness, blastResistance, toolType, toolLevel));
		block.setCreativeTabs(ctab);
		block.setStrength(hardness, blastResistance).setHarvestLevel(toolType, toolLevel);
		block.registerBlock();
		return block;
	}

    public BlockBase(Material material, String intName) {
        super(material);
        this.setBlockTextureName(Reference.MOD_ID + ":" + intName);
        this.setInternalName(intName);
    }
    
    public BlockBase setStrength (float hardness, float blastResistance) {
        this.setHardness(hardness);
        this.setResistance(blastResistance);
        return this;
    }
    
    public BlockBase setCreativeTabs (CreativeTabs creativeTab) {
        this.setCreativeTab(creativeTab);
        return this;
    }
    
    public BlockBase setInternalName(String name){
        this.setBlockName(name);
        return this;
    }
    
    public void registerBlock () {
        GameRegistry.registerBlock(this, Reference.MOD_ID+":"+this.getUnlocalizedName().substring(5));
    }
    
    public void setHarvestLevel(String toolType, int toolLevel){
        super.setHarvestLevel(toolType, toolLevel);
    }

    public void onBlockPlacedBy (World world, int x, int y, int z, EntityLivingBase player, ItemStack stack) {
        int dir = MathHelper.floor_double((double)((player.rotationYaw * 4F) / 360F) + 0.5D) & 3;
        world.setBlockMetadataWithNotify(x, y, z, dir, 0);
    }

}

File: src/main/scala/geek/lawsof/physics/lib/objects/block/BlockBase.scala
package geek.lawsof.physics.lib.objects.block

import net.minecraft.block.material.Material
import net.minecraft.block.Block
import net.minecraft.creativetab.CreativeTabs
import cpw.mods.fml.common.registry.GameRegistry
import net.minecraft.world.World
import net.minecraft.entity.EntityLivingBase
import net.minecraft.item.ItemStack
import net.minecraft.util.{IIcon, AxisAlignedBB, MathHelper}
import geek.lawsof.physics.init.ModConfigs
import geek.lawsof.physics.lib.objects.info.BlockInfo
import geek.lawsof.physics.lib.objects.CreativeTabBase
import net.minecraft.tileentity.TileEntity
import geek.lawsof.physics.lib.objects.block.te.ICustomRenderedBlock
import geek.lawsof.physics.Reference
import net.minecraft.client.renderer.texture.IIconRegister

/**
 * Created by anshuman on 26-05-2014.
 */
class BlockBase(blockMaterial: Material, val intName: String) extends Block(blockMaterial) {
  this.setInternalName(intName)

  def this(intName: String, blockMat: Material, ctab: CreativeTabs, hardness: Float, blastResistance: Float, toolType: String, toolLevel: Int) {
    this(blockMat, intName)
    setCreativeTabs(ctab)
    setStrength(hardness, blastResistance).setHarvestLevel(toolType, toolLevel)
    registerBlock()
  }

  def this(intName: String, blockMat: Material, ctab: CreativeTabBase, info: BlockInfo) {
    this(blockMat, intName)
    setCreativeTabs(ctab)
    setStrength(info.hardness, info.blastResistance).setHarvestLevel(info.toolType, info.toolLevel)
    registerBlock()
  }

  def setStrength(hardness: Float, blastResistance: Float): BlockBase = {
    this.setHardness(hardness)
    this.setResistance(blastResistance)
    this
  }

  def setCreativeTabs(creativeTab: CreativeTabs): BlockBase = {
    this.setCreativeTab(creativeTab)
    this
  }

  def setInternalName(name: String): BlockBase = {
    this.setBlockName(name)
    this
  }

  def registerBlock() {
    GameRegistry.registerBlock(this, this.getUnlocalizedName.substring(5))
  }

  override def setHarvestLevel(toolType: String, toolLevel: Int) {
    super.setHarvestLevel(toolType, toolLevel)
  }

  override def onBlockPlacedBy(world: World, x: Int, y: Int, z: Int, player: EntityLivingBase, stack: ItemStack) {
    val dir: Int = MathHelper.floor_double(((player.rotationYaw * 4F) / 360F).asInstanceOf[Double] + 0.5D) & 3
    world.setBlockMetadataWithNotify(x, y, z, dir, 0)
  }

  def newItemStack(size: Int = 1) = new ItemStack(this, size)

  override def registerBlockIcons(reg : IIconRegister): Unit = this.blockIcon = reg.registerIcon(Reference.MOD_ID + ":" + intName)

  override def getIcon(side : Int, meta : Int): IIcon = this.blockIcon
}
